File: Rokurics-HarmonyOS/entry/.preview/default/cache/default/default@PreviewArkTS/esmodule/debug/entry/src/main/ets/models/ChatModels.ts
/**
 * Chat models
 */
export enum ChatMessageRole {
    SYSTEM = "system",
    USER = "user",
    ASSISTANT = "assistant"
}
export class ChatMessage {
    id: string = '';
    role: ChatMessageRole = ChatMessageRole.USER;
    content: string = '';
    createdAt: Date = new Date();
    attachmentIDs: string[] = [];
    constructor(role: ChatMessageRole, content: string) {
        this.id = ChatMessage.generateId();
        this.role = role;
        this.content = content;
        this.createdAt = new Date();
        this.attachmentIDs = [];
    }
    static generateId(): string {
        const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        let id = '';
        for (let i = 0; i < 36; i++) {
            if (i === 8 || i === 13 || i === 18 || i === 23) {
                id += '-';
            }
            else if (i === 14) {
                id += '4';
            }
            else {
                const idx: number = Math.floor(Math.random() * chars.length);
                id += chars.charAt(idx);
            }
        }
        return id;
    }
}
export interface ChatContextFields {
    title: string;
    itemCount: number;
    items: ChatContextItem[];
    maxContextCharacters: number;
    id?: string;
    browsePathComponents?: string[];
    sourceKind?: string | null;
    sourceItemID?: string | null;
    contextPathDisplay?: string | null;
    itemTitle?: string | null;
    totalCharacterCount?: number;
    isTruncated?: boolean;
    createdAt?: Date;
}
export class ChatContext {
    id: string = '';
    title: string = '';
    browsePathComponents: string[] = [];
    itemCount: number = 0;
    items: ChatContextItem[] = [];
    sourceKind: string | null = null;
    sourceItemID: string | null = null;
    contextPathDisplay: string | null = null;
    itemTitle: string | null = null;
    createdAt: Date = new Date();
    maxContextCharacters: number = 0;
    totalCharacterCount: number = 0;
    isTruncated: boolean = false;
    static create(fields: ChatContextFields): ChatContext {
        const ctx = new ChatContext();
        ctx.id = fields.id ?? ChatMessage.generateId();
        ctx.title = fields.title;
        ctx.browsePathComponents = fields.browsePathComponents ?? [];
        ctx.itemCount = fields.itemCount;
        ctx.items = fields.items;
        ctx.sourceKind = fields.sourceKind ?? null;
        ctx.sourceItemID = fields.sourceItemID ?? null;
        ctx.contextPathDisplay = fields.contextPathDisplay ?? null;
        ctx.itemTitle = fields.itemTitle ?? null;
        ctx.createdAt = fields.createdAt ?? new Date();
        ctx.maxContextCharacters = fields.maxContextCharacters;
        ctx.totalCharacterCount = fields.totalCharacterCount ?? fields.items.reduce((sum, i) => sum + i.content.length, 0);
        ctx.isTruncated = fields.isTruncated ?? false;
        return ctx;
    }
}
export class ChatContextItem {
    id: string = '';
    title: string = '';
    filingPath: string = '';
    content: string = '';
    sourcePath: string | null = null;
    contentCharacterCount: number = 0;
    isTruncated: boolean = false;
    constructor(id: string, title: string, filingPath: string, content: string) {
        this.id = id;
        this.title = title || '未命名知识';
        this.filingPath = filingPath;
        this.content = content.trim();
        this.contentCharacterCount = this.content.length;
        this.isTruncated = false;
    }
}
export class ChatConversation {
    id: string = '';
    title: string = '新对话';
    messages: ChatMessage[] = [];
    activeContextID: string | null = null;
    createdAt: Date = new Date();
    updatedAt: Date = new Date();
    constructor(title?: string) {
        this.id = ChatMessage.generateId();
        this.title = title ?? '新对话';
        this.messages = [];
        this.activeContextID = null;
        this.createdAt = new Date();
        this.updatedAt = new Date();
    }
}
export class ChatRequest {
    messages: ChatMessage[] = [];
    context: ChatContext | null = null;
    modelName: string | null = null;
    maxTokens: number = 2000;
    temperature: number = 0.3;
}
export class ChatResult {
    message: ChatMessage;
    providerID: string = '';
    providerName: string = '';
    modelName: string | null = null;
    finishReason: string | null = null;
    outputWasTruncated: boolean = false;
    constructor(message: ChatMessage, providerID: string, providerName: string) {
        this.message = message;
        this.providerID = providerID;
        this.providerName = providerName;
        this.modelName = null;
        this.finishReason = null;
        this.outputWasTruncated = false;
    }
}

File: Rokurics-Windows/Rokurics/Models/ChatModels.cs
namespace Rokurics.Models;

/// <summary>
/// Chat message, conversation, context models. Mirrors ChatModels from source.
/// </summary>

public enum ChatMessageRole
{
    System,
    User,
    Assistant
}

public sealed class ChatMessage
{
    public string Id { get; set; }
    public ChatMessageRole Role { get; set; }
    public string Content { get; set; }
    public DateTime CreatedAt { get; set; }
    public List<string> AttachmentIds { get; set; }

    public ChatMessage()
    {
        Id = Guid.NewGuid().ToString();
        Content = "";
        AttachmentIds = new List<string>();
    }

    public ChatMessage(string id, ChatMessageRole role, string content, DateTime? createdAt = null, List<string>? attachmentIds = null)
    {
        Id = id;
        Role = role;
        Content = content;
        CreatedAt = createdAt ?? DateTime.UtcNow;
        AttachmentIds = attachmentIds ?? new List<string>();
    }
}

public enum ChatConversationTitleSource
{
    Manual,
    AiGenerated,
    Fallback
}

public enum ChatAttachmentKind
{
    Image,
    Document,
    Audio,
    Other
}

public sealed class ChatAttachment
{
    public string Id { get; set; }
    public string ConversationId { get; set; }
    public string FileName { get; set; }
    public string FileType { get; set; }
    public string? MimeType { get; set; }
    public string RelativePath { get; set; }
    public long SizeBytes { get; set; }
    public DateTime CreatedAt { get; set; }
    public ChatAttachmentKind Kind { get; set; }

    public ChatAttachment()
    {
        Id = Guid.NewGuid().ToString();
        ConversationId = "";
        FileName = "attachment";
        FileType = "";
        RelativePath = "";
    }
}

public sealed class ChatConversation
{
    public string Id { get; set; }
    public string Title { get; set; }
    public DateTime? TitleGeneratedAt { get; set; }
    public ChatConversationTitleSource TitleSource { get; set; }
    public List<ChatMessage> Messages { get; set; }
    public string? ActiveContextId { get; set; }
    public string? ContextPathDisplay { get; set; }
    public int? ContextItemCount { get; set; }
    public string? LastMessagePreview { get; set; }
    public List<string> AttachmentIds { get; set; }
    public List<ChatAttachment> Attachments { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }

    public bool IsEmpty => Messages.Count == 0 && Attachments.Count == 0;

    public ChatConversation()
    {
        Id = Guid.NewGuid().ToString();
        Title = "新对话";
        TitleSource = ChatConversationTitleSource.Fallback;
        Messages = new List<ChatMessage>();
        AttachmentIds = new List<string>();
        Attachments = new List<ChatAttachment>();
        CreatedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public string DisplayTitle => Title;

    public ChatConversation WithMessage(ChatMessage message)
    {
        var conv = Clone();
        conv.Messages = new List<ChatMessage>(Messages) { message };
        conv.UpdatedAt = DateTime.UtcNow;
        if (conv.Messages.Count == 1)
            conv.LastMessagePreview = Truncate(message.Content, 80);
        return conv;
    }

    private ChatConversation Clone()
    {
        return new ChatConversation
        {
            Id = Id, Title = Title, TitleGeneratedAt = TitleGeneratedAt, TitleSource = TitleSource,
            Messages = new List<ChatMessage>(Messages), ActiveContextId = ActiveContextId,
            ContextPathDisplay = ContextPathDisplay, ContextItemCount = ContextItemCount,
            LastMessagePreview = LastMessagePreview, AttachmentIds = new List<string>(AttachmentIds),
            Attachments = new List<ChatAttachment>(Attachments), CreatedAt = CreatedAt, UpdatedAt = UpdatedAt
        };
    }

    private static string Truncate(string text, int maxLen) =>
        text.Length <= maxLen ? text : text[..(maxLen - 3)] + "...";
}

public sealed class ChatContext
{
    public string Id { get; set; }
    public string Title { get; set; }
    public List<string> BrowsePathComponents { get; set; }
    public int ItemCount { get; set; }
    public List<ChatContextItem> Items { get; set; }
    public string? ContextPathDisplay { get; set; }
    public int MaxContextCharacters { get; set; }
    public int TotalCharacterCount { get; set; }
    public bool IsTruncated { get; set; }

    public ChatContext()
    {
        Id = Guid.NewGuid().ToString();
        Title = "学习库";
        BrowsePathComponents = new List<string>();
        Items = new List<ChatContextItem>();
        MaxContextCharacters = 20000;
    }

    public string DisplayTitle => Title;

    public string PathDisplay
    {
        get
        {
            if (!string.IsNullOrWhiteSpace(ContextPathDisplay)) return ContextPathDisplay;
            var components = new List<string> { "学习库" };
            components.AddRange(BrowsePathComponents.Where(c => !string.IsNullOrWhiteSpace(c)).Select(c => c.Trim()));
            return string.Join(" / ", components);
        }
    }
}

public sealed class ChatContextItem
{
    public string Id { get; set; }
    public string Title { get; set; }
    public string FilingPathDisplay { get; set; }
    public string Content { get; set; }
    public string? SourcePath { get; set; }
    public int ContentCharacterCount { get; set; }
    public bool IsTruncated { get; set; }

    public ChatContextItem()
    {
        Id = "";
        Title = "未命名知识";
        FilingPathDisplay = "";
        Content = "";
    }

    public ChatContextItem(string id, string title, string filingPathDisplay, string content, string? sourcePath = null, bool isTruncated = false)
    {
        Id = id;
        Title = string.IsNullOrWhiteSpace(title) ? "未命名知识" : title;
        FilingPathDisplay = filingPathDisplay;
        Content = content.Trim();
        SourcePath = sourcePath?.Trim() is { Length: > 0 } s ? s : null;
        ContentCharacterCount = Content.Length;
        IsTruncated = isTruncated;
    }
}
